### ISSUE TITLE
Annotating and ordering by daily/weekly/monthly stats

### ISSUE THREAD

**Issue by kronok** [M1hacka/django-pg-hll#5]
I'm trying to leverage the power of hll by having annotations for stats within a date range, but unsure of the best way to approach this. Basically, I'm trying to somewhat mimic the date stuff they've described in https://github.com/citusdata/postgresql-hll#data-warehouse-use-case

My use case is needing **all** of the results, but ordered by view count (like needing to order by top viewed this month). So just .filter() within a date range wouldn't work by itself since it filters out objects from the queryset.

Here's a model example:
```
class Thing(models.Model):
    ...


class Stats(models.Model):
    parent = models.ForeignKey(Thing, related_name='stats', on_delete=models.CASCADE)
    hll = HllField()
    date = models.DateField(auto_now_add=True, editable=False)
```

My attempt at it:
```
from django.utils.datetime_safe import date

monthly = date.today() - timedelta(days=30)

queryset = Thing.objects.all().annotate(
    month_view_count=Coalesce(UnionAggCardinality('stats__hll', filter=Q(stats__date__gte=monthly)), 0),
).order_by('month_view_count')

queryset[0]

django.db.utils.ProgrammingError: FILTER specified, but hll_cardinality is not an aggregate function
LINE 1: ...l_timestamp", "things_thing"."featured", COALESCE(hll_cardin...
                                                             ^
```

Any ideas of what I should look more into or try out? I've been trying to avoid getting really messy, and it feels like there's a 1 liner here that I'm not thinking of that would do it.

I'll be looking into postgres conditionals next, but curious if this is easier than what I'm making it out to be.

**Comment by M1hacka** [comment]
Hi.
I'm not sure, if I understood correctly what you want.
Do you want to count cardinality for every month and then sort result by this counter?
Or may be count a floating 30-day window for each day?
From your example:
1. If I undrstood task correctly, it doesn't work. `monthly` will always return date in 30 days from now, it doesn't group results per month.
2. It's quit better to avoid `Thing` and `Stats` join, if you have no need in it. This can be achived with using `Stats` as based models and calling `values` before `annotate(...)`. See the documentation [here](https://docs.djangoproject.com/en/3.0/topics/db/aggregation/#values)

You can try the following (though, I haven't tested it):
1. Count month value for each `Stats` item: `annotate(month=ExtractMonth('date'))`
2. Group results by month: `values('month')`
3. Count hll cardinality for each group: `annotate(month_view_count=Coalesce(UnionAggCardinality('hll'))`
4. Order the result with `month_view_count`

```python
from django.db.models.functions import ExtractMonth, Coalesce
from django_pg_hll.aggregate import UnionAggCardinality
queryset = Stats.objects.annotate(month=ExtractMonth('date')).values('month').annotate(month_view_count=Coalesce(UnionAggCardinality('hll')).order_by('month_view_count')
```

**Comment by kronok** [comment]
I'm looking to do more of a floating 30-day window for each day, but that example you showed is actually really cool and I may use that in the future. I wouldn't have thought of that. Thanks so much for that.

It's more in the realm of having a list of 1,000 Thing objects and needing to order them by the total amount of views in the last x days depending on what the user wants to see (like they want to see all Thing objects for their current query that have the most views in last 7 days for instance). It helps them order by what's popular in the last 7 days, but the Thing objects with 0 views will still show up at the bottom of the list.

So after seeing your example and thinking about it a bit, I felt like I need to do a Stat subquery for my Thing objects.

I started playing around by doing this to see if I could get the right data:
```
from datetime import timedelta
from django.utils.datetime_safe import date
from django.db.models.functions import Coalesce
from django_pg_hll.aggregate import UnionAggCardinality

>>> week_ago = date.today() - timedelta(days=7)
>>> Stats.objects.filter(parent_id=1487, date__gte=week_ago).values('parent_id').annotate(card=Coalesce(UnionAggCardinality('hll'), 0))
<BulkUpdateQuerySet [{'parent_id': 1487, 'card': 77}]>
```
Which is perfect. I just need to turn it into a Subquery:
```
from django.db.models import OuterRef, Subquery

subquery = Subquery(
    Stats.objects.filter(parent=OuterRef("pk"), date__gte=week_ago).values('parent_id').annotate(
            card=Coalesce(UnionAggCardinality('hll'), 0)
    ).values('card')[:1]
)
queryset = Thing.objects.all().annotate(date_view_count=subquery).order_by(F('date_view_count').desc(nulls_last=True))
```
Sounds like it would work, but I run into this:
```
django.core.exceptions.FieldError: Expression contains mixed types. You must set output_field.
```
This is where things get interesting. If I remove the `date__gte=week_ago` from the subquery filter, it runs fine:
```
subquery = Subquery(
    Stats.basic_manager.filter(parent=OuterRef("pk")).values('parent_id').annotate(
            card=Coalesce(UnionAggCardinality('hll'), 0)
    ).values('card')[:1]
)
queryset = Thing.objects.all().annotate(date_view_count=subquery).order_by(F('date_view_count').desc(nulls_last=True))
```
With my current understanding, that makes no sense to me. My current thought is to try wrapping some stuff around things to try to force an output_field to IntegerField.

If you have any clever ways to accomplish this in a better way, I'm all ears too!

**Comment by M1hacka** [comment]
Ok, I've understood your task.

The problem in your first attempt in `filter` expression is that `hll_cardinality` as well as `coalesce` is not an aggregate function. Here `hll_union_agg` is aggregate function, building hll from multiple rows. So you need to build filtered hll query first, and then count `hll_cardinality` over results.

If I use raw SQL, I'd write query you want to achive as:
```sql
SELECT "tests_thing".*, COALESCE(hll_cardinality("week_view_hll"), 0) AS "week_view_count"
FROM (
    SELECT "parent_id", hll_union_agg("tests_stats"."hll")
        FILTER (WHERE "tests_stats"."date" >= NOW() - INTERVAL '7 days') AS "week_view_hll"
    FROM "tests_stats"
    GROUP BY "parent_id"
) AS hll_agg
    LEFT OUTER JOIN "tests_thing" ON ("tests_thing"."id" = "hll_agg"."parent_id")
ORDER BY "week_view_count" ASC
```

Django (as I know) have no ability to build such a join. You can build inner query, but can't join query results with other table and count `hll_cardinality` over result set. In django you can implement it using a subquery, which is a little bit more crazy looking in SQL (and may be less effective, but I'm not sure about it)

```python
subquery = Subquery(Stats.objects.filter(parent=OuterRef('pk')).
    annotate(week_view_hll=UnionAgg('hll', filter=Q(date__gte=week_ago))).values('week_view_hll'))
queryset = Thing.objects.annotate(week_hll_count=Coalesce(Cardinality(subquery), 0)).\
    values('id', 'week_hll_count').order_by('week_hll_count')
```

It generate something like:
```sql
SELECT "tests_thing"."id", COALESCE(hll_cardinality((
    SELECT hll_union_agg(U0."hll") FILTER (WHERE (U0."date" >= 2019-12-08)) AS "month_view_hll" 
    FROM "tests_stats" U0 
    WHERE U0."parent_id" = "tests_thing"."id" 
    GROUP BY U0."id"
)), 0) AS "month_hll_count" 
FROM "tests_thing" 
GROUP BY "tests_thing"."id" 
ORDER BY "month_hll_count" ASC
```

**Comment by kronok** [comment]
Awesome, that was just enough kick to get it working the way I need it to (I think)!

I didn't go for the raw sql approach because this code is in the middle of a django-filter and I'll have to fuss with it a bit and read more docs to see if I can get it working that way because I never work in raw sql. I'd definitely love to do that to speed this up in the future.

So I went for the subquery route. I had to add in the values('parent_id') to make it group everything into 1 value since subqueries require that it only returns 1 value, and your example was returning all of them separately, so you'd only get 1 day of data in the subquery.

Here's what I have that works:
```
week_ago = date.today() - timedelta(days=7)
subquery = Subquery(Stats.objects.filter(parent=OuterRef('pk')).values('parent_id').annotate(view_hll=CardinalitySum('hll', filter=Q(date__gte=week_ago))).values('view_hll')[:1])
qs = Thing.objects.all().annotate(date_view_count=subquery).order_by(F('date_view_count').desc(nulls_last=True))
```
Added CardinalitySum because I count 1 view per day, and this seems to work to where if the same person visits the same Thing multiple times each day for 7 days, it counts as 7 total instead of just 1. (well, I hope that's how it works)

Anyways, I felt so close to making it work the whole time, but you helped me finally get there. Thanks so much and I hope this helps others out there trying to do the same thing!
